controller/controller_server/sync_lock.py

```diff
@@ -0,0 +1,50 @@
+import threading
+from collections import defaultdict
+
+from controller.common.csi_logger import get_stdout_logger
+from controller.controller_server.errors import ObjectAlreadyProcessingError
+
+logger = get_stdout_logger()
+
+ids_in_use = defaultdict(set)
+
+
+def add_to_ids_in_use(lock_key, object_id):
+    ids_in_use[lock_key].add(object_id)
+
+
+class SyncLock:
+    def __init__(self, lock_key, object_id, action_name):
+        self._lock = threading.Lock()
+        self.lock_key = lock_key
+        self.object_id = object_id
+        self.action_name = action_name
+
+    def __enter__(self):
+        if self.lock_key:

```

Comment:
why these `if`s? could `self` be uninitialized here?

Reply:
`lock_key` can be `None` for methods that need the handle_exception but do not operate on volumes

Reply:
I think you mean it can be `''` (and `object_id` can be `None`). I see.
it's weird to use `SyncLock` even when not actually needing it, but it does simplify the `call_csi_method` logic.
this makes me think about some unexpected usages of `SyncLock` - could we change `_remove_from_ids_in_use` to something like:
```
    if object_id in ids_in_use[lock_key]:
        ids_in_use[lock_key].remove(object_id)
    else:
        logger.error("could not find lock to release for action {} with {}: {}".format(action_name, lock_key, lock_id))
```
just in case?

Reply:
for any non-`''` value?

Reply:
sorry for the confusion, my suggestion is about different scenarios than those related to the beginning of this conversation.
i.e. it's in addition to the current `if self.lock_key:`

Reply:
done